### ISSUE TITLE
False positive - callable result of an overloaded method is reported to be not callable

### ISSUE THREAD

**Issue by PatrickBourdon** [PyCQA/pylint#4660]
### Steps to reproduce
```python
from typing import Union, Any, Literal, overload
from collections.abc import Callable


def my_print(*args: Any) -> None:
    print(", ".join(str(x) for x in args))
    return


# ---- This is OK ----
class MyClass:
    def my_method(self, option: Literal["mandatory"]) -> Callable[..., Any]:
        return my_print


c = MyClass().my_method("mandatory")
c(1, "foo")

# ---- This runs OK but pylint reports an error ----
class MyClass1:
    @overload
    def my_method(self, option: Literal["mandatory"]) -> Callable[..., Any]:
        ...

    @overload
    def my_method(
        self, option: Literal["optional", "mandatory"]
    ) -> Union[None, Callable[..., Any]]:
        ...

    def my_method(
        self, option: Literal["optional", "mandatory"]
    ) -> Union[None, Callable[..., Any]]:
        return my_print


d = MyClass1().my_method("mandatory")
d(1, "bar")  # => d is not callable - pylint(not callable)

```

### Current behavior
Runs ok.
But pylint reports an error on the last line of code: d is not callable - pylint(not-callable)

This appeared on some existing code after updating to pylint 2.9.0.
I reduced to the smallest piece of code to reproduce the issue.
Note that It does not appear when there is no overload, since 'c' is Ok.

### Expected behavior
No pylint error report.

### Environment
Windows 10
pylint 2.9.3
astroid 2.6.2
Python 3.9.6 (tags/v3.9.6:db3ff76, Jun 28 2021, 15:26:21) [MSC v.1929 64 bit (AMD64)]

VScode 1.57.1
python extension: v2021.6.944021595
pylance: v2021.6.3

Thank you for your great job and tool.
Patrick, Paris, France,
